registry/consul/registry.go

```diff
@@ -0,0 +1,166 @@
+package consul
+
+import (
+	"context"
+	"fmt"
+	"slices"
+	"sync"
+	"time"
+
+	consul "github.com/hashicorp/consul/api"
+	"github.com/moderntv/cadre/registry"
+	"google.golang.org/grpc/grpclog"
+)
+
+var logger = grpclog.Component("consul_registry")
+
+var _ registry.Registry = &consulRegistry{}
+
+type consulRegistry struct {
+	client        *consul.Client
+	datacenter    string
+	refreshPeriod time.Duration
+	aliases       map[string]string
+	mu            sync.RWMutex
+	services      map[string][]registry.Instance
+}
+
+func NewRegistry(address, datacenter string, aliases map[string]string, refreshPeriod time.Duration) (registry.Registry, error) {
+	config := consul.DefaultConfig()
+	config.Address = address
+	c, err := consul.NewClient(config)
+	if err != nil {
+		return nil, err
+	}
+
+	r := &consulRegistry{
+		client:        c,
+		datacenter:    datacenter,
+		refreshPeriod: refreshPeriod,
+		aliases:       aliases,
+		services:      make(map[string][]registry.Instance),
+	}
+	return r, nil
+}
+
+func (r *consulRegistry) Register(serviceInstance registry.Instance) error {
+	return nil
+}
+
+func (r *consulRegistry) Deregister(serviceInstance registry.Instance) error {
+	return nil
+}
+
+func (r *consulRegistry) Instances(service string) []registry.Instance {
+	r.mu.RLock()
+	defer r.mu.RUnlock()
+	is, ok := r.services[service]
+	if !ok {
+		return []registry.Instance{}
+	}
+
+	return is
+}
+
+func (r *consulRegistry) Watch(service string) (<-chan registry.RegistryChange, func()) {
+	ch := make(chan registry.RegistryChange)
+	ctx, cancel := context.WithCancel(context.Background())
+	go r.watch(ctx, service, ch)
+	closeFn := func() {
+		close(ch)
+		cancel()
+	}
+	return ch, closeFn
+}
+
+func (r *consulRegistry) watch(ctx context.Context, service string, ch chan<- registry.RegistryChange) {
+	var consulService string
+	alias, ok := r.aliases[service]
+	if ok {
+		consulService = alias
+		logger.Infof("[CONSUL REGISTRY] using alias (%s) to resolve service (%s)", alias, service)
+	}
+
+	if !ok {
+		consulService = service
+		logger.Warningf("[CONSUL REGISTRY] no alias defined for service (%s)", service)
+	}
+
+	logger.Infof("[CONSUL REGISTRY] watching changes for service (%s) every (%s)", service, r.refreshPeriod)
+	ticker := time.NewTicker(r.refreshPeriod)
+	for {
+		r.resolveService(service, consulService, ch)
+		logger.Infof("[CONSUL REGISTRY] checked changes for service (%s)", service)
+		select {
+		case <-ticker.C:
+
+		case <-ctx.Done():
+			logger.Infof("[CONSUL REGISTRY] canceled watch for service (%s)", service)
+			return
+		}
+	}
+}
+
+func (r *consulRegistry) resolveService(service, consulService string, ch chan<- registry.RegistryChange) {
+	q := &consul.QueryOptions{
+		Datacenter: r.datacenter,
+	}
```

Comment:
Add health check for QueryOptions or use new endpoint

Reply:
Will be changed as agreed.